rfcs/62-editor-data-model.md

```diff
@@ -0,0 +1,555 @@
+
+This RFC describes the data model and API that the Editor uses in memory, and the basis for communicating with the Game process and for serializing that data to disk (although the serialization format to disk is outside of the scope of this RFC). The data model is largely inspired by the one of [Our Machinery](https://ourmachinery.com/), as described in [this blog post](https://ourmachinery.com/post/the-story-behind-the-truth-designing-a-data-model/).
+
+## Glossary
+
+In this RFC, the following terms are used:
+
+- **Game** refers to any application made with Bevy, which encapsulates proper games but also any other kind of Bevy application (CAD, architecture software, _etc._).
+- **Author** refers to any person (including non-technical) building the Game with the intent of shipping it to end users.
+- **Editor** refers to the software used to edit Game data and prepare them for runtime consumption, and which this RFC describes the data model of.
+- **Runtime** refers to the Game or Editor process running time, as opposed to compiling or building time. When not otherwise specified, this refers to the Game's runtime. In general it's synonymouns of a process running, and is opposed to _build time_ or _compile time_.
+
+## Motivation
+
+We need a data model for the Bevy Editor. We want to make it decoupled from the runtime data model that the Game uses, which is based on compiled Rust types (`World`, _etc._) with a defined native layout from a fixed Rust compiler version (as most types don't use `#[repr(C)]`), so we can handle data migration across Rust compiler versions and therefore across both Editor and Game versions. We also want to make it _transportable_, that is serializable with the intent to transport it over any IPC mechanism, so that the Editor process can dialog with the Game process to enable real-time Game editing ("Play Mode").
+
+## User-facing explanation
+
+The _data model_ of the Editor refers to the way the Editor manipulates editing data, their representation in memory while the Editor executable is running, and the API to do those manipulations. The Editor at its core defines a centralized data model, and all systems use the same API to access that data model, making it the unique source of truth while the Editor process is running. This unicity prevents desynchronization between various subsystems, since they share a same view of the editing data. It also enables global support for common operations such as copy/paste and undo/redo without requiring each subsystem to implement their own variant.
+
+A Bevy application (the Game) uses Rust types defined in one of the `bevy_*` crates, any other third-party plugin library, or the Game itself (_custom types_). The data is represented in memory at runtime by _instances_ (allocated objects) of those Rust types. This representation is optimal in terms of read and write access, but is _static_, that is the type layout is defined during compiling and cannot be changed afterward.

```

Comment:
> You missed the types in the std library and types in non-plugin library.

Again, I've never said or written that all Rust types are concerned by the IPC mechanism. I very much doubt the Editor cares about `Option<T>` as a type; we probably want to limit to components, resources, events, and a few others relevant for editing.

I'm not sure what you mean by "non-plugin library" though.

> Also I'm questioning if this sentence is relevant to the RFC.

This shows the contrast between the Game data model and the Editor one, and why we cannot just have a single data model for both.

Reply:
> Again, I've never said or written that all Rust types are concerned by the IPC mechanism.

Oops, I thought you tried to be exhaustive when enumerating the types that the Game was aware about. It looks like you're just listing what Game types need to be shared (through reflection) with the Editor.

> I'm not sure what you mean by "non-plugin library" though.

I was meaning a dependency of the Game that is not a Bevy plugin, e.g. the `rand` crate. But again, probably not something we want the Editor to care about.

> This shows the contrast between the Game data model and the Editor one, and why we cannot just have a single data model for both.

I think there is a more explicit and clear way to expose that. It probably should be mentioned in the beginning of the paragraph, such as “The Editor and Game data models are of different format because [...]. Therefore the two must exchange Game type data through an IPC mechanism”, or something among those lines. It is much more clear IMO. But I think that for it to work, the “A Bevy application (the Game) uses Rust types [...]” sentence should be simplified. In particular, I think there is no need to distinguish the origin of the types. It does not care (at least to a certain level) if they are defined inside the Project's crate/workspace or defined in a dependency. And *I think*, reflection data contains namespace info up to the crate name if we really need that info, and we get it for free.

Reply:
> The Editor and Game data models are of different format because [...]. Therefore the two must exchange Game type data through an IPC mechanism.

Nit-picking a bit here, but technically different formats is not a cause that implies IPC as a consequence. IPC is only for separate processes. Although I understand what you mean, _assuming_ separate processes _then_ IPC is a consequence of the formats being different.

Reply:
Applied some of those suggestions, breaking up the user-facing explanation into sections and defining some missing terms.